ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Monitor.hs

```diff
@@ -571,16 +578,36 @@ monitorLedgerStateJudgement :: ( MonadSTM m
                             => PeerSelectionActions peeraddr peerconn m
                             -> PeerSelectionState peeraddr peerconn
                             -> Guarded (STM m) (TimedDecision m peeraddr peerconn)
-monitorLedgerStateJudgement PeerSelectionActions{ readLedgerStateJudgement }
+monitorLedgerStateJudgement PeerSelectionActions{ readLedgerStateJudgement,
+                                                  peerTargets = ConsensusModePeerTargets{
+                                                    praosTargets,
+                                                    genesisSyncTargets } }
                             st@PeerSelectionState{ bootstrapPeersFlag,
                                                    publicRootPeers,
                                                    knownPeers,
                                                    establishedPeers,
                                                    inProgressPromoteCold,
                                                    inProgressPromoteWarm,
-                                                   ledgerStateJudgement
-                                                 }
-  | isBootstrapPeersEnabled bootstrapPeersFlag =
+                                                   ledgerStateJudgement,
+                                                   consensusMode }
+  | GenesisMode <- consensusMode =
+    Guarded Nothing $ do
+      lsj <- readLedgerStateJudgement
+      check (lsj /= ledgerStateJudgement)
+      let targets = case lsj of
+            YoungEnough -> praosTargets
+            TooOld      -> genesisSyncTargets
+
+      return $ \_now ->
+        Decision {
+          decisionTrace = [TraceLedgerStateJudgementChanged lsj],
+          decisionJobs  = [],
+          decisionState = st {
+              ledgerStateJudgement = lsj,
+              targets } }
+
+  | PraosMode <- consensusMode
+  , isBootstrapPeersEnabled bootstrapPeersFlag =
```

Comment:
This should be safe. The targets updated in here are for the private PeerSelectionState. Thanks to your work on PeerSelectionCounters, churn loop is now updating the TVar targets at one call site, so I've added a check there for the current FetchMode and so this will synchronize the targets basis with the OG before making the update.

Reply:
I've realized there is indeed a bug and made some minor changes to fix it.

Reply:
We should design a test which can catch something like this.

Reply:
This can be outside of this PR, but it would be nice to have a spec on how churn works and test the implementation against the spec.

Reply:
Yes, a spec would be very neat, I'll create an issue for it. The test would fall under the umbrella of the test we should write (and maybe there's an issue for it already) - are the current/effective targets matching the demanded targets.